book/src/asm/instructions.md

```diff
@@ -20,10 +20,50 @@ Instructions feature:
 
 # External instructions
 
-An external instruction delegates calls to a function inside a submachine of this machine. When it is called, a call is made to the submachine function. An example of an external instruction is the following:
+An external instruction delegates its implementation to a function/operation from a submachine.
+When called, the inputs and outputs of the declared instruction are mapped into a call to the submachine function.
 
+Assume we have a submachine with a single operation `add`:
 ```
-instr assert_zero X = my_submachine.assert_zero // where `assert_zero` is a function defined in `my_submachine`
+machine SubMachine(latch, operation_id) {
+    operation add<0> a,b -> c;
+    ...
+}
+```
+
+In the simplest case, the parameters of an instruction match one to one, in order, with the call to the target operation:
+```
+machine Main {
+    ...
+    reg X[<=];
+    reg Y[<=];
+    reg Z[<=];
+
+    instr add X, Y -> Z = submachine.add; // will call submachine.add(X, Y) and return through Z
+    // this is equivalent to:
+    // instr add X, Y -> Z = submachine.add X, Y -> Z;
+    ...
+}
+```
+
+The mapping is flexible and allows using, besides the assignment registers in the instruction parameters, any regular registers from the local machine (i.e., the caller).
+All the following mappings are possible:
+```
+machine Main {
+    reg X[<=];
+    reg Y[<=];
+    reg Z[<=];
+    reg A;
+    reg B;
+    reg C;
+    ...
+    instr add_to_A X, Y = addvm.add X, Y -> A;          // - output written to a regular register
+    instr addAB -> X = addvm.add A, B -> X;             // - input from the value of regular registers
+    instr addAB_to_C = addvm.add A, B -> C;             // - regular registers used for both inputs and output
+    instr addAB_to_A = addvm.add A, B -> A;             // - read from and then write to the same regular register
+    instr sub_from_add X, Y -> Z = addvm.add Y, Z -> X; // - find input for a given output (doesn't always work!)
+    ...
+}
 ```
```

Comment:
I think we should also mention somewhere that assignment registers have to be "repeated" on the lhs while non-assignment registers can only appear on the rhs. Not sure how to say that in an understandable way, though :)